src/lib/upgradePackageDefinitions.ts

```diff
@@ -49,18 +127,47 @@ export async function upgradePackageDefinitions(
 
   if (options.peer && Object.keys(filteredLatestDependencies).length > 0) {
     const upgradedPeerDependencies = await getPeerDependenciesFromRegistry(filteredLatestDependencies, options)
-    const peerDependencies = { ...options.peerDependencies, ...upgradedPeerDependencies }
-    if (!dequal(options.peerDependencies, peerDependencies)) {
-      const [newUpgradedDependencies, newLatestVersions, newPeerDependencies] = await upgradePackageDefinitions(
-        { ...currentDependencies, ...filteredUpgradedDependencies },
-        { ...options, peerDependencies, loglevel: 'silent' },
-      )
-      return [
-        { ...filteredUpgradedDependencies, ...newUpgradedDependencies },
-        { ...latestVersionResults, ...newLatestVersions },
-        newPeerDependencies,
-      ]
+
+    let rerunResult: RerunUpgradeResult
+    let checkPeerViolationResult: CheckIfInPeerViolationResult = {
+      violated: false,
+      filteredUpgradedDependencies,
+      upgradedPeerDependencies,
     }
+    let runIndex = 0
+    do {
+      rerunResult = await rerunUpgradeIfChangedPeers(
+        currentDependencies,
+        checkPeerViolationResult.filteredUpgradedDependencies,
+        checkPeerViolationResult.upgradedPeerDependencies,
+        latestVersionResults,
+        options,
+      )
+      if (!rerunResult) {
+        if (runIndex > 0) {
+          // We can't find anything to do, will not upgrade anything
+          return [{}, latestVersionResults, options.peerDependencies]
+        }
+        checkPeerViolationResult = checkIfInPeerViolation(
+          currentDependencies,
+          filteredUpgradedDependencies,
+          upgradedPeerDependencies,
+        )
+        if (!checkPeerViolationResult.violated) {
+          // No issues were found, return
+          return [filteredUpgradedDependencies, latestVersionResults, options.peerDependencies]
+        }
```

Comment:
This if statement can be removed once the loop condition changes to `while (checkPeerViolationResult.violated)`.

Reply:
Done